bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java

```diff
@@ -74,10 +83,121 @@
+        private final LedgerMetadataIndex ledgerIndex;
+
+        /**
+         * Construct an Watchable with zero watchers.
+         */
+        public TransientLedgerInfo(long ledgerId, LedgerMetadataIndex ledgerIndex) {
+            super(WATCHER_RECYCLER);
+            this.ledgerId = ledgerId;
+            this.ledgerIndex = ledgerIndex;
+        }
+
+        synchronized Long getLastAddConfirmed() {
+            return lac;
+        }
+
+        Long setLastAddConfirmed(long lac) {
+            long lacToReturn;
+            boolean changed = false;
+            synchronized (this) {
+                if (null == this.lac || this.lac < lac) {
+                    this.lac = lac;
+                    changed = true;
+                }
+                lacToReturn = this.lac;
+            }
+            if (changed) {
+                notifyWatchers(lacToReturn);
+            }
+            return lacToReturn;
+        }
+
+        synchronized boolean waitForLastAddConfirmedUpdate(long previousLAC,
+                                                           Watcher<LastAddConfirmedUpdateNotification> watcher)
+                throws IOException {
+            if ((null != lac && lac > previousLAC)
+                    || isClosed || ledgerIndex.get(ledgerId).getFenced()) {
+                return false;
+            }
+
+            addWatcher(watcher);
+            return true;
+        }
+
+        public ByteBuf getExplicitLac() {
+            ByteBuf retLac = null;
+            synchronized (this) {
+                if (explicitLac != null) {
+                    retLac = Unpooled.buffer(explicitLac.capacity());
+                    explicitLac.rewind(); //copy from the beginning
+                    retLac.writeBytes(explicitLac);
+                    explicitLac.rewind();
+                    return retLac;
+                }
+            }
+            return retLac;
+        }
+
+        public void setExplicitLac(ByteBuf lac) {
+            long explicitLacValue;
+            synchronized (this) {
+                if (explicitLac == null) {
+                    explicitLac = ByteBuffer.allocate(lac.capacity());
+                }
+                lac.readBytes(explicitLac);
+                explicitLac.rewind();
+
+                // skip the ledger id
+                explicitLac.getLong();
+                explicitLacValue = explicitLac.getLong();
+                explicitLac.rewind();
+            }
+            setLastAddConfirmed(explicitLacValue);
+        }
+
+        void notifyWatchers(long lastAddConfirmed) {
+            notifyWatchers(LastAddConfirmedUpdateNotification.FUNC, lastAddConfirmed);
+        }
+
+        @Override
+        public void close() {
+           synchronized (this) {
+               if (isClosed) {
+                   return;
+               }
+               isClosed = true;
+           }
+           // notify watchers
+            notifyWatchers(Long.MAX_VALUE);
```

Comment:
watchers are being told that the lac for the ledger is MAX_VALUE? Seems like this will violate TOAB.

Reply:
we don't use the value notified as the source-of-truth. the value is used for trigger the logic when the lac is changed. we re-read lac and put that value in the response. we did this because LAC can potentially be updated between the watcher received notification and sending the response.  re-read lac would give the most recent value, so the client can have as many entries to read before hitting lac again.

Reply:
ok. seems a bit ropey, since MAX_VALUE is a valid value for lac in theory, and someone could try to add an optimization in future and end up sending it to the client, but that's not for this patch to deal with anyhow.

Reply:
"MAX_VALUE" can't really be a valid value for lac. what would be the entry id to carry "MAX_VALUE" lac?

Reply:
not valid read from storage, no. But a watcher could receive a positive long from the Watchable, and conceivably consider it valid.

Reply:
> But a watcher could receive a positive long from the Watchable, and conceivably consider it valid.

yes the only case for Long.MAX_VALUE is from watchers. Then the meaning given for Long.MAX_VALUE is the watcher logic. At this case, Long.MAX_VALUE would then be effectively same as using some negative values.

Reply:
Which means the watcher will need to know that MAX_VALUE is a special value. It's a implicit contract.

Reply:
yes that's the contract.